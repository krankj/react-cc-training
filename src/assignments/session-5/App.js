import React from "react";
import ArrayList from "./ArrayList";
import GroceryList from "./GroceryList";

const App = (props) => <GroceryList />;

export default App;
