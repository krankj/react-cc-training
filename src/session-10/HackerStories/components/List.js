import styles from "./List.module.css";

function List({ stories, handleRemoveStory }) {
  return (
    <div>
      <ol>
        {stories.map(function (item, index) {
          return (
            <li key={index}>
              <span className={styles.item}>
                <a href={item.url}>{item.title}</a>
              </span>
              <span className={styles.item}>{item.author}</span>
              <span className={styles.item}>{item.num_comments}</span>
              <span className={styles.item}>{item.points}</span>
              <span>
                <button type="button" onClick={() => handleRemoveStory(item)}>
                  Dismiss
                </button>
              </span>
            </li>
          );
        })}
      </ol>
    </div>
  );
}
export default List;
