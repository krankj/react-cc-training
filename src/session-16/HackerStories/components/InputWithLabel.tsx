import { useEffect, useRef, useState } from "react";

type InputWithLabelProps = {
  id: string;
  value: string;
  type?: string;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  isFocused?: boolean;
  children?: React.ReactNode;
};

const InputWithLabel = ({
  id,
  children,
  type = "text",
  value,
  onInputChange,
  isFocused,
}: InputWithLabelProps) => {
  const [clickAutoFocus, setClickAutoFocus] = useState(false);

  const inputRef = useRef<HTMLInputElement>(null!);
  useEffect(() => {
    if (inputRef.current && clickAutoFocus) {
      inputRef.current.focus();
    }
  }, [clickAutoFocus]);

  return (
    <>
      <label htmlFor={id}>Search: </label>
      <input
        style={{
          display: "block",
          padding: "0.5rem",
          borderRadius: "24px",
          border: "none",
          outline: "none",
          width: "350px",
        }}
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
      />
    </>
  );
};

export default InputWithLabel;
