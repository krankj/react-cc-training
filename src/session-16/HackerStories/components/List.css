.item {
  display: inline-block;
  width: 200px;
  padding: 1rem;
  color: wheat;
}

.title {
  color: ghostwhite;
  text-decoration: none;
}

.button {
  border: none;
  padding: 0.5rem;
  background-color: rgb(185, 1, 1);
  color: whitesmoke;
  border-radius: 6px;
}

.button:hover {
  cursor: pointer;
}
