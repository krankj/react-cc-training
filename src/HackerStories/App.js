import List from "./components/List";
import Search from "./components/Search";
import styles from "./App.module.css";
import StateForSearch from "./components/StateForSearch";
import { useState } from "react";

const stories = [
  {
    title: "React",
    url: "https://reactjs.org/",
    author: "Jordan Walke",
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: "Redux",
    url: "https://redux.js.org/",
    author: "Dan Abramov, Andrew Clark",
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
];

function App() {
  const [searchTerm, setSearchTerm] = useState("");
  const handleOnSearch = (e) => {
    setSearchTerm(e.target.value);
  };

  const filteredStories = stories.filter((story) =>
    story.title.toLowerCase().includes(searchTerm.toLowerCase())
  );

  return (
    <div className={styles.container}>
      <Search onSearch={handleOnSearch} />
      <h4>Searching for: {searchTerm} </h4>
      <List stories={filteredStories} />
    </div>
  );
}
export default App;
