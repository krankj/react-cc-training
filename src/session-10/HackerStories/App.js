import List from "./components/List";
import Search from "./components/Search";
import styles from "./App.module.css";
import StateForSearch from "./components/StateForSearch";
import React, { useState, useEffect } from "react";
import usePersistenceState from "./hooks/usePersistenceState";
import InputWithLabel from "./components/InputWithLabel";

const initStories = [
  {
    title: "React",
    url: "https://reactjs.org/",
    author: "Jordan Walke",
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: "Redux",
    url: "https://redux.js.org/",
    author: "Dan Abramov, Andrew Clark",
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
  {
    title: "Karma",
    url: "https://reactjs.org/",
    author: "Sadhguru",
    num_comments: 3,
    points: 4,
    objectID: 2,
  },
  {
    title: "In Search Of Lost Time",
    url: "https://redux.js.org/",
    author: "Marcel Proust",
    num_comments: 2,
    points: 5,
    objectID: 3,
  },
  {
    title: "Ulysses",
    url: "https://reactjs.org/",
    author: "James Joyce",
    num_comments: 3,
    points: 4,
    objectID: 4,
  },
  {
    title: "100 years of solitude",
    url: "https://redux.js.org/",
    author: "Gabriel Garcia",
    num_comments: 2,
    points: 5,
    objectID: 5,
  },
];

function App() {
  const [searchTerm, setSearchTerm] = usePersistenceState(
    "customSearchTerm",
    ""
  );
  const [stories, setStories] = useState(initStories);
  const handleOnSearch = (e) => {
    setSearchTerm(e.target.value);
  };

  const handleRemoveStory = (item) => {
    const newStories = stories.filter(
      (story) => item.objectID !== story.objectID
    );
    setStories(newStories);
  };
  const filteredStories = stories.filter((story) =>
    story.title.toLowerCase().includes(searchTerm.toLowerCase())
  );

  return (
    <div className={styles.container}>
      <InputWithLabel
        id="search"
        value={searchTerm}
        onInputChange={handleOnSearch}
        type="text"
      >
        <strong>Search For Stories</strong>
      </InputWithLabel>
      <h4>Searching for: {searchTerm} </h4>
      <List stories={filteredStories} handleRemoveStory={handleRemoveStory} />
    </div>
  );
}
export default App;
