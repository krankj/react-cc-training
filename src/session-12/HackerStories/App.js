import List from "./components/List";
import Search from "./components/Search";
import styles from "./App.module.css";
import StateForSearch from "./components/StateForSearch";
import React, { useState, useEffect, useReducer } from "react";
import usePersistenceState from "./hooks/usePersistenceState";
import InputWithLabel from "./components/InputWithLabel";

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

const storiesReducer = (state, action) => {
  switch (action.type) {
    case "STORIES_FETCH_INIT":
      return { ...state, isLoading: true, isError: false };
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        data: action.payload,
        isLoading: false,
        isError: false,
      };
    case "STORIES_FETCH_FAILURE":
      return { ...state, data: [], isLoading: false, isError: true };
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter((i) => action.payload.objectID !== i.objectID),
      };
    default:
      throw new Error();
  }
};

function App() {
  const [searchTerm, setSearchTerm] = usePersistenceState(
    "customSearchTerm",
    ""
  );

  const [stories, dispatchStories] = useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  useEffect(() => {
    dispatchStories({ type: "STORIES_FETCH_INIT" });
    fetch(`${API_ENDPOINT}react`)
      .then((res) => res.json())
      .then((result) => {
        dispatchStories({
          type: "STORIES_FETCH_SUCCESS",
          payload: result.hits,
        });
      })
      .catch(() => {
        dispatchStories({ type: "STORIES_FETCH_FAILURE" });
      });
  }, []);

  const handleOnSearch = (e) => {
    setSearchTerm(e.target.value);
  };

  const onHandleDeleteItem = (item) => {
    dispatchStories({ type: "REMOVE_STORY", payload: item });
  };

  const filteredStories = stories.data.filter((story) =>
    story.title.toLowerCase().includes(searchTerm.toLowerCase())
  );

  return (
    <div className={styles.container}>
      <InputWithLabel
        id="search"
        value={searchTerm}
        onInputChange={handleOnSearch}
        type="text"
      >
        <strong>Search For Stories</strong>
      </InputWithLabel>
      <h4>Searching for: {searchTerm} </h4>
      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List stories={filteredStories} handleDeleteItem={onHandleDeleteItem} />
      )}
      {stories.isError && <p>Something went wrong..</p>}
    </div>
  );
}
export default App;
