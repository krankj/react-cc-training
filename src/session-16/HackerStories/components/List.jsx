import { memo, useContext } from "react";
import { LabelContext } from "../App";

import "./List.css";

const List = memo(({ stories, handleDeleteItem }) => {
  return (
    <div>
      <ol>
        {stories.map(function (item, index) {
          return (
            <Item key={index} item={item} handleDeleteItem={handleDeleteItem} />
          );
        })}
      </ol>
    </div>
  );
});

const Item = ({ item, handleDeleteItem }) => (
  <li>
    <span className="item">
      <a className="title" href={item.url}>
        {item.title}
      </a>
    </span>
    <span className="item">{item.author}</span>
    <span className="item">{item.num_comments}</span>
    <span className="item">{item.points}</span>
    <span>
      <DeleteButton onClickHandler={() => handleDeleteItem(item)} />
    </span>
  </li>
);

const DeleteButton = ({ label, onClickHandler }) => {
  return (
    <button className="button button_small" onClick={onClickHandler}>
      Delete
    </button>
  );
};

export default List;

export { Item };
