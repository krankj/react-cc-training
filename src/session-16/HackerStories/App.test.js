import App, { storiesReducer, List, SearchForm } from "./App";
import InputWithLabel from "./components/InputWithLabel";
import { Item } from "./components/List";
import { render, screen, fireEvent, act } from "@testing-library/react";
import axios from "axios";

//  1:
// describe("dummy suite", () => {
//   test("true to be true", () => {
//     expect(true).toBe(true);
//   });
//   test("false to be false", () => {
//     expect(false).toBe(false);
//   });
// });

// test("true to be true", () => {
//   expect(true).toBe(true);
// });
// test("false to be false", () => {
//   expect(false).toBe(false);
// });

//Large subjects like functions or components often require multiple test cases, so it makes sense to
//use them with test suites:
// describe("App component", () => {
//   test("removes an item when clicking the Dismiss button", () => {});
//   test("requests some initial stories from an API", () => {});
// });

//“test” block can also be written as an “it” block
// describe("something truthy and falsy", () => {
//   it("true to be true", () => {
//     expect(true).toBe(true);
//   });
//   it("false to be false", () => {
//     expect(false).toBe(false);
//   });
// });

// failing a test
// describe("something truthy and falsy", () => {
//   test("true to be true", () => {
//     expect(true).toBe(true);
//   });
//   test("false to be false", () => {
//     expect(false).toBe(true);
//   });
// });

//Unit Testing: Functions
//Let's start with testing basic functions
// Let us define some test data

const storyOne = {
  title: "React",
  url: "https://reactjs.org/",
  author: "Tommy Hill",
  num_comments: 3,
  points: 4,
  objectID: 0,
};

const storyTwo = {
  title: "Redux",
  url: "https://redux.js.org/",
  author: "Dan Abramov, Andrew Clark",
  num_comments: 2,
  points: 5,
  objectID: 1,
};
const stories = [storyOne, storyTwo];

// describe("storiesReducer", () => {
//   test("removes a story from all stories", () => {});
// });

//Note: This test will fail. Why?
// describe("storiesReducer", () => {
//   test("removes a story from all stories", () => {
//     const action = { type: "REMOVE_STORY", payload: storyOne };
//     const state = { data: stories, isLoading: false, isError: false };
//     const newState = storiesReducer(state, action);
//     const expectedState = {
//       data: [storyTwo],
//       isLoading: false,
//       isError: false,
//     };
//     expect(newState).toStrictEqual(expectedState);
//   });
// });

//Unit Testing: Components

// describe("Item", () => {
//   test("renders all properties", () => {
//     render(<Item item={storyOne} />);
//   });
// });

//We can use debug to see the rendered output

// describe("Item", () => {
//   test("renders all properties", () => {
//     render(<Item item={storyOne} />);
//     screen.debug();
//   });
// });

// describe("Item", () => {
//   test("renders all properties", () => {
//     render(<Item item={storyOne} />);
//     expect(screen.getByText("Tommy Hill")).toBeInTheDocument();
//     expect(screen.getByText("React")).toHaveAttribute(
//       "href",
//       "https://reactjs.org/"
//     );
//   });
// });

// describe("Item", () => {
//   test("renders all properties", () => {
//     render(<Item item={storyOne} />);
//     expect(screen.getByText("Tommy Hill")).toBeInTheDocument();
//     expect(screen.getByText("React")).toHaveAttribute(
//       "href",
//       "https://reactjs.org/"
//     );
//   });
//   test("renders a clickable delete button", () => {
//     render(<Item item={storyOne} />);
//     expect(screen.getByRole("button")).toBeInTheDocument();
//   });
// });

// describe("Item", () => {
//   test("renders all properties", () => {
//     render(<Item item={storyOne} />);
//     expect(screen.getByText("Tommy Hill")).toBeInTheDocument();
//     expect(screen.getByText("React")).toHaveAttribute(
//       "href",
//       "https://reactjs.org/"
//     );
//   });
//   test("renders a clickable dismiss button", () => {
//     render(<Item item={storyOne} />);
//     expect(screen.getByRole("button")).toBeInTheDocument();
//   });

//   test("clicking the dismiss button calls the callback handler", () => {
//     const handleRemoveItem = jest.fn();
//     render(<Item item={storyOne} handleDeleteItem={handleRemoveItem} />);
//     fireEvent.click(screen.getByRole("button"));
//     expect(handleRemoveItem).toHaveBeenCalledTimes(1);
//   });
// });

//Testing the searchFrom

// describe("SearchForm", () => {
//   const searchFormProps = {
//     searchTerm: "React",
//     onSearchInput: jest.fn(),
//     onSearchSubmit: jest.fn(),
//   };
//   test("renders the input field with its value", () => {
//     render(<SearchForm {...searchFormProps} />);
//     screen.debug();
//   });
// });

// describe("SearchForm", () => {
//   const searchFormProps = {
//     searchTerm: "React",
//     onSearchInput: jest.fn(),
//     onSearchSubmit: jest.fn(),
//   };
//   test("renders the input field with its value", () => {
//     render(<SearchForm {...searchFormProps} />);
//     expect(screen.getByDisplayValue("React")).toBeInTheDocument();
//   });
// });

//This will fail since we don't have label yet
// describe("SearchForm", () => {
//   const searchFormProps = {
//     searchTerm: "React",
//     onSearchInput: jest.fn(),
//     onSearchSubmit: jest.fn(),
//   };
//   test("renders the input field with its value", () => {
//     render(<SearchForm {...searchFormProps} />);
//     expect(screen.getByDisplayValue("React")).toBeInTheDocument();
//   });

//   test("renders the correct label", () => {
//     render(<SearchForm {...searchFormProps} />);
//     expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
//   });
// });

// describe("SearchForm", () => {
//   const searchFormProps = {
//     searchTerm: "React",
//     onSearchInput: jest.fn(),
//     onSearchSubmit: jest.fn(),
//   };
//   test("renders the input field with its value", () => {
//     render(<SearchForm {...searchFormProps} />);
//     expect(screen.getByDisplayValue("React")).toBeInTheDocument();
//   });

//   test("calls onSearchInput on input field change", () => {
//     render(<SearchForm {...searchFormProps} />);
//     fireEvent.change(screen.getByDisplayValue("React"), {
//       target: { value: "Redux" },
//     });
//     expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
//   });
//   test("calls onSearchSubmit on button submit click", () => {
//     render(<SearchForm {...searchFormProps} />);
//     fireEvent.submit(screen.getByRole("button"));
//     expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
//   });
// });

//Integration Testing: Component

jest.mock("axios");

// describe("App", () => {
//   test("succeeds fetching data", () => {
//     const promise = Promise.resolve({
//       data: {
//         hits: stories,
//       },
//     });
//     axios.get.mockImplementationOnce(() => promise);
//     render(<App />);
//     screen.debug();
//   });
// });

// describe("App", () => {
//   test("succeeds fetching data", async () => {
//     const promise = Promise.resolve({
//       data: {
//         hits: stories,
//       },
//     });
//     axios.get.mockImplementationOnce(() => promise);
//     render(<App />);
//     screen.debug();
//     await act(() => promise);
//     screen.debug();
//   });
// });

// describe("App", () => {
//   test("succeeds fetching data", async () => {
//     const promise = Promise.resolve({
//       data: {
//         hits: stories,
//       },
//     });
//     axios.get.mockImplementationOnce(() => promise);
//     render(<App />);
//     expect(screen.queryByText(/Loading/)).toBeInTheDocument();
//     await act(() => promise);
//     expect(screen.queryByText(/Loading/)).toBeNull();
//   });
// });

// describe("App", () => {
//   test("succeeds fetching data", async () => {
//     const promise = Promise.resolve({
//       data: {
//         hits: stories,
//       },
//     });
//     axios.get.mockImplementationOnce(() => promise);
//     render(<App />);
//     expect(screen.queryByText(/Loading/)).toBeInTheDocument();
//     await act(() => promise);
//     expect(screen.queryByText(/Loading/)).toBeNull();
//     expect(screen.getByText("React")).toBeInTheDocument();
//     expect(screen.getByText("Redux")).toBeInTheDocument();
//     expect(screen.getAllByText("Delete").length).toBe(2);
//   });
// });

// describe("App", () => {
//   test("fails fetching data", async () => {
//     const promise = Promise.reject();
//     axios.get.mockImplementationOnce(() => promise);
//     render(<App />);
//     expect(screen.getByText(/Loading/)).toBeInTheDocument();
//     try {
//       await act(() => promise);
//     } catch (error) {
//       expect(screen.queryByText(/Loading/)).toBeNull();
//       expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
//     }
//   });
//   test("removes a story", async () => {
//     const promise = Promise.resolve({
//       data: {
//         hits: stories,
//       },
//     });
//     axios.get.mockImplementationOnce(() => promise);
//     render(<App />);
//     await act(() => promise);
//     expect(screen.getAllByText("Delete").length).toBe(2);
//     expect(screen.getByText("Tommy Hill")).toBeInTheDocument();
//     fireEvent.click(screen.getAllByText("Delete")[0]);
//     expect(screen.getAllByText("Delete").length).toBe(1);
//     expect(screen.queryByText("Tommy Hill")).toBeNull();
//   });
// });

//Snapshot Testing:

describe("SearchForm", () => {
  const searchFormProps = {
    searchTerm: "React",
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn(),
  };
  test("renders the input field with its value", () => {
    render(<SearchForm {...searchFormProps} />);
    screen.debug();
  });

  test("renders snapshot", () => {
    const { container } = render(<SearchForm {...searchFormProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
