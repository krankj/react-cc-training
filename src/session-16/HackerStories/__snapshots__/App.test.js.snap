// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders snapshot 1`] = `
<form
  class="searchForm"
>
  <label
    for="search"
  >
    Search: 
  </label>
  <input
    id="search"
    style="display: block; padding: 0.5rem; border-radius: 24px; outline: none; width: 350px;"
    type="text"
    value="React"
  />
  <button
    class="button_submit"
    type="submit"
  >
    Submit
  </button>
</form>
`;
